### PR TITLE
OAK-10870 : New oak-mongo.js methods for sizes of properties, counts …

### PR DESCRIPTION
**Pull request by stefan-egli**
…by clusterId and unmerged branch commits

### CODE REVIEW

File: oak-run/src/main/js/oak-mongo.js [hunk 1]
```diff
@@ -128,6 +128,28 @@ var oak = (function(global){
...
+     */
+    api.countDirectChildren = function(path){
+        if (path === undefined) {
+            return 0;
+        } else if (path != "/") {
+            path = path + "/";
+        }
+        var depth = pathDepth(path);
+        var totalCount = 0;
+        var count = db.nodes.count({_id: pathFilter(depth++, path)});
```
Comment (shodaaan) [inline oak-run/src/main/js/oak-mongo.js]: Using depth + 1 instead of depth++ might be a bit more clear.

File: oak-run/src/main/js/oak-mongo.js [hunk 2]
```diff
@@ -573,6 +595,201 @@ var oak = (function(global){
...
     };
 
+    /**
+     * Prtints the sizes of all revisions by property.
+     * This is useful for large documents to quickly see which property/ies are affected
+     *
+     * @memberof oak
+     * @method propertySizes
+     * @param {string} path the path of a document
+     * @param {number} sizeLargerThan only show properteis larger than this, defaults to 1
```
Comment (shodaaan) [inline oak-run/src/main/js/oak-mongo.js]: small typo: properteis

File: oak-run/src/main/js/oak-mongo.js [hunk 3]
```diff
@@ -573,6 +595,201 @@ var oak = (function(global){
...
+        }
+        print("loading document at " + path);
+        var doc = this.findOne(path);
+        if (!doc) {
+            print("No document for path: " + path);
+            return;
+        }
+        var overall = Object.bsonsize(doc);
+        print("overall size : " + overall);
+        var k;
```
Comment (shodaaan) [inline oak-run/src/main/js/oak-mongo.js]: Renaming k to "prop" might make this a bit more clear.

File: oak-run/src/main/js/oak-mongo.js [hunk 4]
```diff
@@ -573,6 +595,201 @@ var oak = (function(global){
...
+            if (subdocsize <= sizeLargerThan) {
+                //print("   (too small to report " + k + ")");
+                continue;
+            }
+            print(" - property " + k + " size : " + subdocsize);
+        }
+    }
+
+    /**
+     * Prtints the count of property revisions by clusterId
```
Comment (shodaaan) [inline oak-run/src/main/js/oak-mongo.js]: Prtints - should this be "Prints" or is it something else?

File: oak-run/src/main/js/oak-mongo.js [hunk 5]
```diff
@@ -573,6 +595,201 @@ var oak = (function(global){
...
+        }
+        print("loading document at " + path);
+        var doc = this.findOne(path);
+        if (!doc) {
+            print("No document for path: " + path);
+            return;
+        }
+        var k;
+        var stats = {};
+        for (k in doc) {
```
Comment (shodaaan) [inline oak-run/src/main/js/oak-mongo.js]: Same here: "prop" instead of "k" would make it more clear.

File: oak-run/src/main/js/oak-mongo.js [hunk 6]
```diff
@@ -573,6 +595,201 @@ var oak = (function(global){
...
+            return;
+        }
+        print("loading document at " + path);
+        var doc = this.findOne(path);
+        if (!doc) {
+            print("No document for path: " + path);
+            return;
+        }
+        var num = 0;
+        var r;
```
Comment (shodaaan) [inline oak-run/src/main/js/oak-mongo.js]: I'm assuming "r" here is "revision": renaming to "rev" or adding a comment would make this a bit more clear.

File: oak-run/src/main/js/oak-mongo.js [hunk 7]
```diff
@@ -573,6 +595,201 @@ var oak = (function(global){
...
+                print(" - branch change " + r + " is committed");
+                continue;
+            }
+            print(" - branch change " + r + " is not or not yet committed");
+            num++;
+        }
+        print("Number of unmerged branch changes : " + num);
+    }
+
+    api.sizeOf = function(obj) {
```
Comment (shodaaan) [inline oak-run/src/main/js/oak-mongo.js:744]: Please consider adding a short description for what this method does.

File: oak-run/src/main/js/oak-mongo.js [hunk 8]
```diff
@@ -573,6 +595,201 @@ var oak = (function(global){
...
+        } else if (thetype == "[object Null]") {
+            return 0;
+        } else if (thetype == "[object String]") {
+            return obj.length;
+        }
+        print("sizeOf: obj not a bson but : " + thetype);
+        return 42;
+    }
+
+    api.unmergedBranchStatsForProperty = function(path, property) {
```
Comment (shodaaan) [inline oak-run/src/main/js/oak-mongo.js:768]: Please consider adding a short description for what this method does.

File: oak-run/src/main/js/oak-mongo.js [hunk 9]
```diff
@@ -573,6 +595,201 @@ var oak = (function(global){
...
+        if (!doc) {
+            print("No document for path: " + path);
+            return;
+        }
+        var subdoc = doc[property];
+        var num = 0;
+        var totalGarbage  = 0;
+        var cachedRoot = this.findOne("/");
+        var propertyGarbageSizes = {};
+        var r, clusterId, clusterIdGarbage;
```
Comment (shodaaan) [inline oak-run/src/main/js/oak-mongo.js]: Renaming "r" to "rev" would make it more clear.

File: oak-run/src/main/js/oak-mongo.js (hunk 4 above)
Comment (stefan-egli) [reply oak-run/src/main/js/oak-mongo.js]: should be prints

**Review by Joscorbe — APPROVED** [review]
Looks good

**Comment by stefan-egli** [comment]
thx @shodaaan for the comments - I have addressed all of them in e401911223ddc78c1958ba11d8d2cf6c724b085d - please review, thx

**Review by shodaaan — APPROVED** [review]

**Review by shodaaan — APPROVED** [review]
Approved, thank you!

**Pull request merged**